compiler/rustc_mir_dataflow/src/impls/live_borrows.rs

```diff
@@ -0,0 +1,642 @@
+
+                let node_kind = match assign_place.ty(self.local_decls, self.tcx).ty.kind() {
+                    ty::Ref(..) | ty::RawPtr(..) => NodeKind::Borrow(assign_local),
+                    _ => NodeKind::LocalOrLocalWithRefs(assign_local),
+                };
+
+                let node_idx = self.dep_graph.add_node(node_kind);
+                self.locals_to_node_indexes.insert(assign_local, node_idx);
+                self.current_local = Some(node_idx);
+
+                debug!(
+                    "set current_local to {:?} for local {:?}",
+                    self.current_local, assign_local
+                );
+
+                // Do not call `visit_place` here as this might introduce a self-edge, which our liveness analysis
+                // assumes not to exist.
+                self.visit_rvalue(&assign.1, location)
+            }
+            StatementKind::FakeRead(..)
+            | StatementKind::StorageDead(_)
+            | StatementKind::StorageLive(_) => {}
+            _ => {
+                self.super_statement(statement, location);
+            }
+        }
+
+        self.current_local = None;
+    }
+
+    #[instrument(skip(self), level = "debug")]
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        debug!("self.current_local: {:?}", self.current_local);
+        match rvalue {
+            Rvalue::Ref(_, _, borrowed_place) | Rvalue::AddressOf(_, borrowed_place) => {
+                let Some(src_node_idx) = self.current_local else {
+                    bug!("Expected self.current_local to be set with Rvalue::Ref|Rvalue::AddressOf");
+                };
+
+                let borrowed_local = borrowed_place.local;
+                let node_idx =
+                    if let Some(node_idx) = self.locals_to_node_indexes.get(&borrowed_local) {
+                        *node_idx
+                    } else {
+                        self.dep_graph.add_node(NodeKind::Borrow(borrowed_place.local))
+                    };
+
+                debug!(
+                    "adding edge from {:?}({:?}) -> {:?}({:?})",
+                    src_node_idx,
+                    self.dep_graph.node(src_node_idx).data,
+                    node_idx,
+                    self.dep_graph.node(node_idx).data,
+                );
+
+                self.dep_graph.add_edge(src_node_idx, node_idx, ());
+            }
+            _ => self.super_rvalue(rvalue, location),
+        }
+    }
+
+    #[instrument(skip(self), level = "debug")]
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        // Add edges for places that correspond to references or raw pointers
+        let place_ty = place.ty(self.local_decls, self.tcx).ty;
+        debug!(?place_ty);
+        debug!("current local: {:?}", self.current_local);
+        match place_ty.kind() {
+            ty::Ref(..) | ty::RawPtr(..) => match self.current_local {
```

Comment:
IIUC, this assumes that we can only store a pointer in a variable which has somewhere type `Ref` or `RawPtr`. What about a `usize` storing a pointer whose address has been exposed?

Reply:
I hadn't thought about this before implementing. This does make things more complicated and requires us to be more conservative. I have considered this in the current implementation, though haven't fully eliminated unsafety possibilities yet (see the FIXMEs in the module doc-comment).